examples/data_preprocessing/police_killing/data_preprocessing_with_keywords.ipynb

```diff
@@ -0,0 +1,1274 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mapping_rules_labels_t = get_mapping_rules_labels_t(rule2label, num_classes)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Build the Z matrix (instances x rules)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Get the Train Data\n",
+    "\n",
+    "For now, the word pairs contained in the *rule2rule_id* dictionary are simple strings. To use them as actual rules, we should convert them to regexes firstly to be able to look for the exact words in the word pair. For example, the words in the rules must not end in another word character (a-z). If the killing word in a rule is, for instance, \\\"kill\\\", only sentences containing the word \\\"kill\\\" should be matched and no sentences containing words like \\\"kills\\\" or \\\"killed\\\" because there are separate rules for these words. (This is ensured by appending \"\\W\" - the word not followed by any letter - to both the police and the killing word contained in a rule.) Technically, it is realized by creating a dictionary containing the strings of word pairs as keys and a list of the corresponding RegEx as values.\n",
+    "\n",
+    "Secondly, we apply the regexes to the samples. In the beginning, everything is stored in a list of dictionaries (one for each individual sample in the train data). The dictionaries contain the sample, a list of the rules matched in it, and a list of the corresponding rule IDs. After that, we convert the list of dictionaries to a Pandas Dataframe."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def convert_rules2regex(rule2rule_id: Dict) -> Dict:\n",
+    "       \n",
+    "        return {rule: [f'{rule.split()[0]}\\W', f'{rule.split()[1]}\\W'] for rule in rule2rule_id}\n",
+    "\n",
+    "def get_data_for_df(data: pd.DataFrame, searches: Dict) -> List:\n",
+    "\n",
+    "    data_for_df = []\n",
+    "\n",
+    "    for sample in tqdm(data[\"samples\"].drop_duplicates()):\n",
+    "        data_dict = {}\n",
+    "        data_dict[\"samples\"] = sample\n",
+    "        data_dict[\"rules\"] = []\n",
+    "        data_dict[\"enc_rules\"] = []\n",
+    "   \n",
+    "        for rule, search in searches.items():\n",
+    "            if re.search(search[0], sample.lower()) and re.search(search[1], sample.lower()):\n",
+    "                data_dict[\"rules\"].append(rule)\n",
+    "                data_dict[\"enc_rules\"].append(rule2rule_id[rule])\n",
+    "\n",
+    "        data_for_df.append(data_dict)\n",
+    "\n",
+    "    return data_for_df\n",
+    "\n",
+    "\n",
+    "def get_df(rule2rule_id: Dict, data: pd.DataFrame) -> pd.DataFrame:\n",
```

Comment:
the function names should usually have some meaningful names that reflect what's going on inside the function